pkg/licensing/classifier.go

```
@@ -84,3 +103,226 @@ func Classify(filePath string, r io.Reader, confidenceLevel float64) (*types.Lic
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	return <-p.pool
+}
+
+// puts the classifier back to the pool of classifiers
+func (p *licenseClassifierPool) Put(classifier *classifier.Classifier) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	p.pool <- classifier
+}
+
+type ClassifierInput struct {
+	PkgID               string
+	FilePath            string
+	Content             []byte
+	ConfidenceLevel     float64
+	LicenseTextCacheDir string
+}
+
+// Classify detects and classifies the license found in a file.
+// It also extracts the license text for each match, creates and persist them in in given folder
+// Each file would be of form <license-text-checksum>.txt
+func (input *ClassifierInput) Classify() (*types.LicenseFile, error) {
+	// check if classifierPool is successfully initialized beforehand or not
+	if classifierPool == nil {
+		return nil, errors.New("please initialize license classifier pool before calling Classify method")
+	}
+
+	var findings types.LicenseFindings
+	var matchType types.LicenseType
+	seen := make(map[string]struct{})
+
+	classifier := classifierPool.Get()
+	defer classifierPool.Put(classifier)
+
+	result := classifier.Match(classifier.Normalize(input.Content))
+
+	for _, match := range result.Matches {
+		if match.Confidence <= input.ConfidenceLevel {
+			continue
+		}
+		if _, ok := seen[match.Name]; ok {
+			continue
+		}
+		seen[match.Name] = struct{}{}
+
+		switch match.MatchType {
+		case "Header":
+			matchType = types.LicenseTypeHeader
+		case "License":
+			matchType = types.LicenseTypeFile
+		}
+
+		licenseLink := fmt.Sprintf("https://spdx.org/licenses/%s.html", match.Name)
+
+		// extract & get the checksum and license text for the current match
+		textChecksum, licenseText := extractLicenseText(input.Content, match.StartLine, match.EndLine)
+
+		// persist the license text in given folder
+		if input.LicenseTextCacheDir != "" && len(licenseText) > 0 {
+			go input.persistLicenseText(textChecksum, licenseText)
```

won't this fail silently in case of error and impact scan result?

---

how will it impact scan result?